### PR TITLE
Detailed Card

### PR DESCRIPTION
**Pull request by mashrukh-islam**
### Ticket Link
---------------------------------------------------

### Related Links
---------------------------------------------------

### Description
---------------------------------------------------

### Steps to Reproduce / Test
---------------------------------------------------

### GIF's
---------------------------------------------------
DetailCard

### CODE REVIEW

File: app/containers/ITunesProvider/TrackDetailsContainer/index.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  trackDetails,
+  maxWidth,
+  trackDetailsError,
+  trackId,
+  dispatchClearTrackDetails,
+  dispatchFetchTrackDetails
+}) {
+  const [loading, setLoading] = useState(true);
+  useInjectSaga({ key: 'iTunesProvider', saga });
+  const params = useParams();
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/TrackDetailsContainer/index.js]: destructure here itself
so 
```
const { trackId } = useParams(); 
```

File: app/containers/ITunesProvider/TrackDetailsContainer/index.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            <T id="view_album" />
+          </a>,
+          <a href={trackDetails.trackViewUrl} key="artist">
+            <T id="view_track" />
+          </a>
+        ]}
+      >
+        <Row>
+          <Col size={1}>
+            <Statistic title="Track Name" value={trackDetails.trackCensoredName} />
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/TrackDetailsContainer/index.js]: use react-intl, lets make it a practice. No stray strings

File: app/containers/ITunesProvider/TrackDetailsContainer/index.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      </Card>
+    </Container>
+  );
+}
+
+TrackDetailsContainer.propTypes = {
+  dispatchFetchTrackDetails: PropTypes.func,
+  dispatchClearTrackDetails: PropTypes.func,
+  trackId: PropTypes.string,
+  trackDetails: PropTypes.object,
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/TrackDetailsContainer/index.js]: its not always needed but it'll be good to do it this one time. Lets actually chart out the shape of the object using PropTypes.shape

File: app/containers/ITunesProvider/TracksContainer/index.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+export function TracksContainer({
+  maxWidth,
+  dispatchFetchTracks,
+  dispatchClearTracks,
+  artistName,
+  tracks,
+  tracksError
+}) {
+  useInjectSaga({ key: 'iTunesProvider', saga });
+  const [loading, setLoading] = useState(false);
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/TracksContainer/index.js]: do we need another loading state variable? 

There is a "loader" that we can read from the redux store yes?

File: app/containers/ITunesProvider/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+// import PropTypes from 'prop-types'
+import { connect } from 'react-redux';
+import { injectIntl, FormattedMessage as T } from 'react-intl';
+import { createStructuredSelector } from 'reselect';
+import { compose } from 'redux';
+import { useInjectSaga } from '@utils/injectSaga';
+import { makeSelectITunesProvider } from './selectors';
+import saga from './saga';
+
+export function ITunesProvider() {
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/index.js]: we can delete this right?

File: app/containers/ITunesProvider/saga.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    yield put(successGetTrackNames(data));
+  } else {
+    yield put(failureGetTrackNames(data));
+  }
+}
+
+export function* fetchTrackDetails(action) {
+  const track = yield select(selectTrackById(action.trackId));
+  if (track) {
+    yield put(successGetTrackDetails(track));
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/saga.js:29]: this doesn't really append to the list, so it doesn't handle the case that we discussed today

File: app/containers/ITunesProvider/TrackDetailsContainer/index.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+      </Card>
+    </Container>
+  );
+}
+
+TrackDetailsContainer.propTypes = {
+  dispatchFetchTrackDetails: PropTypes.func,
+  dispatchClearTrackDetails: PropTypes.func,
+  trackDetails: PropTypes.shape({
+    artistId: PropTypes.number,
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/TrackDetailsContainer/index.js:131]: some out of these will be required right?

File: app/containers/ITunesProvider/TracksContainer/index.js (hunk 4 above)
Comment (alichherawalla) [reply app/containers/ITunesProvider/TracksContainer/index.js]: @mashrukh-islam

File: app/containers/ITunesProvider/reducer.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        break;
+      case iTunesProviderTypes.FAILURE_GET_TRACK_NAMES:
+        draft.tracksError = get(action.error, 'message', 'something_went_wrong');
+        break;
+      case iTunesProviderTypes.REQUEST_GET_TRACK_DETAILS:
+        draft.trackId = action.trackId;
+        break;
+      case iTunesProviderTypes.CLEAR_TRACK_DETAILS:
+        return initialState;
+      case iTunesProviderTypes.SUCCESS_GET_TRACK_DETAILS:
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/reducer.js:49]: draft.trackDetails = action.data;
draft.tracks = [...state.tracks, action.data];

File: app/containers/ITunesProvider/saga.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  } else {
+    const response = yield call(getTrackDetails, action.trackId);
+    const { data, ok } = response;
+    if (ok) {
+      yield put(successGetTrackDetails(data.results[0]));
+      yield put(successGetTrackNames(data.results[0]));
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/saga.js]: remove this line

File: app/containers/ITunesProvider/reducer.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+      case iTunesProviderTypes.REQUEST_GET_TRACK_NAMES:
+        draft.artistName = action.artistName;
+        break;
+      case iTunesProviderTypes.CLEAR_TRACK_NAMES:
+        return initialState;
+      case iTunesProviderTypes.SUCCESS_GET_TRACK_NAMES:
+        draft.tracks = action.data;
+        break;
+      case iTunesProviderTypes.FAILURE_GET_TRACK_NAMES:
+        draft.tracksError = get(action.error, 'message', 'something_went_wrong');
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/reducer.js]: get(action.error, 'message', translate('something_went_wrong'));

Comment (alichherawalla) [reply app/containers/ITunesProvider/reducer.js]: move to saga

File: app/containers/ITunesProvider/saga.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import { takeLatest, call, put, select } from 'redux-saga/effects';
+import { iTunesProviderTypes, iTunesProviderCreators } from './reducer';
+import { getTracks, getTrackDetails } from '@services/itunesApi';
+import get from 'lodash/get';
+import { selectTrackById, selectTracksData } from './selectors';
+import { translate } from '@app/components/IntlGlobalProvider/index';
+// Individual exports for testing
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/saga.js]: remove

File: app/containers/ITunesProvider/tests/saga.test.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    };
+    expect(getTracksGenerator.next(apiResponseGenerator(true, tracksResponse)).value).toEqual(
+      put({
+        type: iTunesProviderTypes.SUCCESS_GET_TRACK_NAMES,
+        data: tracksResponse
+      })
+    );
+  });
+
+  it('should ensure that the action SUCCESS_GET_TRACK_DETAILS is dispatched when the store call succeeds', () => {
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/tests/saga.test.js:56]: test both cases here. The one where we are able to fulfil the request from the redux store and the case where we fulfil it from the API

File: app/containers/ITunesProvider/tests/saga.test.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    getTrackDetailsGenerator = fetchTrackDetails({ trackId });
+    const res = getTrackDetailsGenerator.next().value;
+    expect(JSON.stringify(res)).toEqual(JSON.stringify(select(selectTrackById(trackId))));
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/tests/saga.test.js:59]: you don't need to stringify, please go through this article: https://stackoverflow.com/questions/22413009/jasmine-javascript-testing-tobe-vs-toequal

File: app/containers/ITunesProvider/reducer.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+      case iTunesProviderTypes.FAILURE_GET_TRACK_NAMES:
+        draft.tracksError = action.error;
+        break;
+      case iTunesProviderTypes.REQUEST_GET_TRACK_DETAILS:
+        draft.trackId = action.trackId;
+        draft.loading = true;
+        break;
+      case iTunesProviderTypes.CLEAR_TRACK_DETAILS:
+        return initialState;
+      case iTunesProviderTypes.SUCCESS_GET_TRACK_DETAILS:
```
Comment (alichherawalla) [inline app/containers/ITunesProvider/reducer.js:52]: draft.tracks = [....state.tracks, action.data];
 draft.trackDetails = action.data;

File: app/containers/ITunesProvider/tests/saga.test.js (hunk 12 above)
Comment (mashrukh-islam) [reply app/containers/ITunesProvider/tests/saga.test.js:56]: Yea implemented this.

File: app/containers/ITunesProvider/tests/saga.test.js (hunk 13 above)
Comment (mashrukh-islam) [reply app/containers/ITunesProvider/tests/saga.test.js:59]: I tried without stringifying, and I used both toBe and toEqual. But I was getting the same error as before.
`Received: serialiazes to the same string`

File: app/containers/ITunesProvider/reducer.js (hunk 14 above)
Comment (mashrukh-islam) [reply app/containers/ITunesProvider/reducer.js:52]: done

**Pull request merged**